src/pages/posts/comparing-rust-smart-contract-storage-approaches.md

```diff
@@ -0,0 +1,515 @@
+
+And anyhow, this barely counts as an added cost. Any contract of realistic size probably needs some initialization code, anyhow. The only contracts I've seen that don't require initialization are toy contracts like this incrementer.
+
+So this demo contract seems a little more complicated than the NEAR and Soroban versions, because you need to explicitly initialize the contract, setting the `counter` to `0` while you do so. NEAR and Soroban let you default `counter` to `0`, rather than explicitly setting it. But they also both require duplicating contract bytes on-chain over and over. Score one for CosmWasm, imho.
+
+### Errors
+
+One last thing to get out of the way we focus in on storage manipulation.
+
+The CosmWasm template comes with a bunch of error handling stuff. Most notably:
+
+```rust
+#[derive(Error, Debug)]
+pub enum ContractError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+}
+```
+
+You can add your own errors to this enum, and include user-facing messages using [thiserror](https://docs.rs/thiserror/1.0.21/thiserror/).
+
+Then it gets returned throughout calls to the contract in all those `Result`s. For example:
+
+```rust
+pub fn execute(…) -> Result<Response, ContractError> {…}
+```
+
+This is another area where it kinda seems like CosmWasm requires a lot of ceremony, but it's actually just establishing firm conventions (and great usability) for something that all realistic contracts will need. Once you get your bearings, Soroban encourages a [similar approach to CosmWasm](https://soroban.stellar.org/docs/examples/errors), but less baked-in. And NEAR doesn't have any conventions around this; most people [use the `require!` macro](https://docs.near.org/sdk/rust/best-practices) and throw user-facing messages right in the contract logic. Maybe that's fine?
+
+### Getting and setting state!
+
+Finally what we came here to see!
+
+Here's how you define the shape of the state you're going to store:
+
+```rust
+use schemars::JsonSchema;
+use cw_storage_plus::Item;
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct State {
+    pub count: i8,
+}
+
+pub const STATE: Item<State> = Item::new("state");
+```
+
+With a simple `State` struct like this, CosmWasm looks pretty similar to near-sdk-rs. Reminder of what that looked like:
+
+```rust
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::near_bindgen;
+
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+pub struct Contract {
+    counter: i8,
+}
+```
+
+A powerful thing about CosmWasm, though, is that you can store multiple top-level `Item`s. In NEAR, if you want to add more state, you add more keys to that `Contract` struct. This means that any function in NEAR that accesses state needs to pay the gas cost to deserialize the whole object, even if it doesn't use some keys.
```

Comment:
lol or use `riff`s!

Reply:
Do we have a good landing page / README for Riffs yet?